doc/design/api.md

```diff
@@ -0,0 +1,87 @@
+# Design Doc: New Paddle API
+
+To write a Paddle program using the current API, we would have to write two Python source files -- one defines the data provider and the other defines the network or run the for loop.  This doesn't work well with Notebooks.  So we decide to redesign the API.  This document describes the basic design concerns.
+
+## Basic Concepts
+
+The API design depends on basic concepts about deep learning.
+
+### Model
+
+For deep learning, a model includes two parts: the topology and parameters.  Currently, the concept *model* in Paddle contains only the topology, and parameters are in another concept *gradient machine*.  This differs from the intuition and makes it difficult to save/load models.  In this design, we should keep both topology and parameters in a *model*.
```

Comment:
Here is how tensorflow do it for reference:
* Tensorflow has a concept of graph (topology), load graph: https://godoc.org/github.com/tensorflow/tensorflow/tensorflow/go#Graph.Import

* Save and load weights is a Operation on graph, save and load in golang: https://gist.github.com/helinwang/7782c6b2815c334c77653fc0e52b6069